Multi-Dim Buildsystem Support

ax3l opened the pull request:
## Summary

This updates the CMake build system to allow to build multiple AMReX libraries *at once*, by building multiple, uniquely-named `(lib)amrex_[1|2|3]d.[so|a|dll|lib]` artifacts. Developers can request this by specifying a list of dims instead of a single one during configure time, e.g., `-DAMReX_SPACEDIM="1;2;3"`.

Providing this functionality will significantly reduce deployment burdens for:
- conda packages
- spack packages
- Python modules & packages
- probably Julia packages (Johannes, blink twice when you read this)
- HPC modules

because all dimension-related variants can be shipped _at once_ and _do not produce multiple, conflicting packages_ anymore.

Apps can use the same mechanism to pick up the dimensional variants and build multiple artifacts. (This is something we already do, but cannot generalize to package mangers just yet.)

### Backward Compatibility

To stay backwards compatible, the *last* (or traditionally, *only*) passed dimensionality will define a few legacy artifacts:
- [x] the `amrex::amrex`/`amrex` CMake targets
- [x] the `libamrex.[so|a]`/`amrex.[dll|lib]` artifact (symlink)
- [x] the `AMReX_Config.H` file: defaults to the last dim in `AMReX_SPACEDIM` and defines `AMREX_SPACEDIM` via a public CLI define now

## Additional background

I recently wrote an analysis piece lamenting that [compile time software variants that define exclusive features](https://bssw.io/blog_posts/rethinking-software-variants) are a productivity and usability challenge. One particular challenge we face in the shipping of WarpX and pyAMReX is that users (and developers) would like to seamlessly switch between the dims (1-3D) of our codes. Switching between compile-time and runtime description for such elementary workflows can and should be avoided.

Generally, this problem can be solved with more general interfaces in AMReX. Practically, this is a bit of work that we have not yet started.

Weiqun:
> Other than EB and Linear Solvers, most of the amrex stuff built for 3d could work for 1d and 2d, including i/o.

## Things this does not yet solve

As before, this does not solve namespace and/or symbol clashes. We should introduce unique name spaces and rework many globals in AMReX for this, but this is an orthogonal issue.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [x] include documentation in the code and/or rst files, if appropriate

ax3l committed:
Multi-Dim Buildsystem Support

WeiqunZhang commented on Src/EB/CMakeLists.txt:
```diff
@@ -1,105 +1,107 @@
-   AMReX_EBToPVD.cpp
-   )
-endif ()
+    if (AMReX_SPACEDIM EQUAL 3)
```
Should this be
```suggestion
    if (D EQUAL 3)
```

ax3l replied:
Oh, yes! Good catch!

WeiqunZhang commented:
If we build with `-DAMReX_ENABLE_TESTS=ON -DAMReX_SPACEDIM="1;2;3"` and run `ctest`, only a subset of the tests are running. We need to fix this because one of the broken tests is our smoke test for spack and xsdk.

ax3l committed:
Fix EB conditional 3D additions

Co-authored-by: Weiqun Zhang <WeiqunZhang@lbl.gov>

ax3l committed:
CMake Targets and lib: lowercase `d`

ax3l committed:
Tests: Multi-Dim Support

ax3l committed:
Cleaning: Deduplicate Input List

ax3l committed:
Fix Per-Dim Defines

ax3l commented:
Good point, fixed now.

Shall we update a CI test, maybe a fast host-only entry, to build all dims (or at least 1D and 3D) at a time?

ax3l committed:
Tests: Fix Runtime Dir

ax3l committed:
Fix: Include Guard Name Config D

ax3l committed:
test_install: multiple dims

support testing multiple dims at once

WeiqunZhang commented on Docs/sphinx_documentation/source/BuildingAMReX.rst:
```diff
@@ -603,7 +602,13 @@ In the above snippet, ``<amrex-target-name>`` is any of the targets listed in th
+   +-----------------------+-------------------------------------------------+
+   | amrex_2D              |  AMReX library in 2D                            |
+   +-----------------------+-------------------------------------------------+
+   | amrex_3D              |  AMReX library in 3D                            |
```
```suggestion
   | amrex_1d              |  AMReX library in 1D                            |
   +-----------------------+-------------------------------------------------+
   | amrex_2d              |  AMReX library in 2D                            |
   +-----------------------+-------------------------------------------------+
   | amrex_3d              |  AMReX library in 3D                            |
```

WeiqunZhang commented on Docs/sphinx_documentation/source/BuildingAMReX.rst:
```diff
@@ -712,7 +717,7 @@ As an example, consider the following CMake code:
 
     find_package(AMReX REQUIRED 3D EB)
-    target_link_libraries( Foo PUBLIC AMReX::amrex )
+    target_link_libraries(Foo PUBLIC AMReX::amrex_3D)
```
```suggestion
    target_link_libraries(Foo PUBLIC AMReX::amrex_3d)
```

WeiqunZhang commented on Src/Extern/HYPRE/CMakeLists.txt:
```diff
@@ -1,44 +1,46 @@
-target_include_directories( amrex
-   PUBLIC
-   $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}>)
+add_amrex_define(AMREX_USE_HYPRE NO_LEGACY)
```
```suggestion
add_amrex_define(AMREX_USE_HYPRE NO_LEGACY IF AMReX_HYPRE)
```

ax3l replied:
This is already guarded before the include in `Src/CMakeLists.txt`:
```
if (AMReX_HYPRE OR AMReX_PETSC)
   add_subdirectory(Extern/HYPRE)
endif ()
```

There is a comment we need to check applies to the define:
```
    #
    # This file gets processed if either AMReX_PETSC or AMReX_HYPRE are ON
    # If only AMReX_PETSC is ON, we just need to include this directory
    # and a couple of files to the build.
    #
```

ax3l replied:
Hm, maybe we need indeed
```cmake
add_amrex_define(AMREX_USE_HYPRE NO_LEGACY NO_1D IF AMReX_HYPRE)
```
Otherwise PETSc can flip the switch for hypre defines

ax3l replied:
fixed

ax3l committed:
Multi-Dim: HYPRE & PETSc

ax3l committed:
Update Docs: target names

Co-authored-by: Weiqun Zhang <WeiqunZhang@lbl.gov>

JBlaschke commented:
> probably Julia packages (Johannes, blink twice when you read this)

@ax3l blinking...

This makes a lot of sense -- probably will also help Jackie with some of the things she wants to do.

WeiqunZhang commented:
```
cmake_dependent_option( AMReX_EB "Build with Embedded Boundary support" OFF
   "NOT AMReX_SPACEDIM EQUAL 1" OFF )
```
does not work when `AMReX_SPACEDIM` is `1;2;3`. It turns the EB option off. We may have to use `option`.

WeiqunZhang commented:
```
--- a/Tools/CMake/AMReXOptions.cmake
+++ b/Tools/CMake/AMReXOptions.cmake
@@ -273,7 +273,7 @@ option( AMReX_AMRLEVEL  "Build AmrLevel class" ON )
 print_option( AMReX_AMRLEVEL )
 
 cmake_dependent_option( AMReX_EB "Build with Embedded Boundary support" OFF
-   "NOT AMReX_SPACEDIM EQUAL 1" OFF )
+   "2 IN_LIST AMReX_SPACEDIM OR 3 IN_LIST AMReX_SPACEDIM" OFF )
 print_option(AMReX_EB)
 
 cmake_dependent_option( AMReX_FORTRAN_INTERFACES "Build Fortran API" OFF
```
seems to work

ax3l commented:
Thanks for spotting this!

I agree we can make this a regular option now, I added other fail-safes now for EB:
It simply warns and skips for the 1D geometry builds.

ax3l committed:
EB: Always Option

Warns and skips if selected for 1D

ax3l committed:
Merge remote-tracking branch 'mainline/development' into topic-multi-dim-build

ax3l committed:
Test: Parser2 Multi-Dim

ax3l committed:
Start Fixing 2D Hypre Bugs

Likely untested interface changes that need to be updated
in tests and covered in CI for 2D.

WeiqunZhang committed:
Finish fixing Hypre and PETSc for non-MultiFab

ax3l committed:
PETSc: Same Type Update as for HYPRE

ax3l commented on Tools/CMake/AMReXConfig.cmake.in:
```diff
@@ -231,9 +233,17 @@ if (@AMReX_CUDA@)
+list(LENGTH AMReX_SPACEDIM list_len)
+math(EXPR list_last "${list_len} - 1")
+list(GET AMReX_SPACEDIM ${list_last} AMReX_SPACEDIM_LAST)
+add_library(AMReX::amrex ALIAS AMReX::amrex_${AMReX_SPACEDIM_LAST}d)
```
This file seems to be allowed to be included multiple times (seen as a comment in the auto-generated `AMReXTargets.cmake`, which defines the actual ND targets). Thus, we need an include guard for `add_library`:  #3324

ax3l commented on Tools/CMake/AMReXConfig.cmake.in:
```diff
@@ -231,9 +233,17 @@ if (@AMReX_CUDA@)
+if (@CMAKE_VERSION@ VERSION_GREATER_EQUAL 3.20 AND @AMReX_CUDA@)
    # CUDA architectures amrex was built for -- should we make
    set(AMREX_CUDA_ARCHS @AMREX_CUDA_ARCHS@ CACHE INTERNAL "CUDA archs AMReX is built for")
    set_target_properties(AMReX::amrex
```
Ooops, should use the non-alias targets. Fix in #3324
